source/windows/bcrypt_aes.c

```diff
@@ -0,0 +1,1110 @@
+        s_aes_gcm_algorithm_handle,
+        BCRYPT_CHAINING_MODE,
+        (PUCHAR)BCRYPT_CHAIN_MODE_GCM,
+        (ULONG)(wcslen(BCRYPT_CHAIN_MODE_GCM) + 1),
+        0);
+
+    AWS_FATAL_ASSERT(NT_SUCCESS(status) && "BCryptSetProperty for GCM chaining mode failed");
+
+    /* Setup CTR algorithm */
+    status = BCryptOpenAlgorithmProvider(&s_aes_ctr_algorithm_handle, BCRYPT_AES_ALGORITHM, NULL, 0);
+    AWS_ASSERT(s_aes_ctr_algorithm_handle && "BCryptOpenAlgorithmProvider() failed");
+
+    /* This is ECB because windows doesn't do CTR mode for you.
+       Instead we use ECB and XOR the encrypted IV and data to operate on for each block. */
+    status = BCryptSetProperty(
+        s_aes_ctr_algorithm_handle,
+        BCRYPT_CHAINING_MODE,
+        (PUCHAR)BCRYPT_CHAIN_MODE_ECB,
+        (ULONG)(wcslen(BCRYPT_CHAIN_MODE_ECB) + 1),
+        0);
+
+    AWS_FATAL_ASSERT(NT_SUCCESS(status) && "BCryptSetProperty for ECB chaining mode failed");
+
+    /* Setup KEYWRAP algorithm */
+    status = BCryptOpenAlgorithmProvider(&s_aes_keywrap_algorithm_handle, BCRYPT_AES_ALGORITHM, NULL, 0);
+    AWS_ASSERT(s_aes_ctr_algorithm_handle && "BCryptOpenAlgorithmProvider() failed");
+
+    AWS_FATAL_ASSERT(NT_SUCCESS(status) && "BCryptSetProperty for KeyWrap failed");
+}
+
+static BCRYPT_KEY_HANDLE s_import_key_blob(
+    BCRYPT_ALG_HANDLE algHandle,
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *key) {
+    NTSTATUS status = 0;
+
+    BCRYPT_KEY_DATA_BLOB_HEADER key_data;
+    key_data.dwMagic = BCRYPT_KEY_DATA_BLOB_MAGIC;
+    key_data.dwVersion = BCRYPT_KEY_DATA_BLOB_VERSION1;
+    key_data.cbKeyData = (ULONG)key->len;
+
+    struct aws_byte_buf key_data_buf;
+    aws_byte_buf_init(&key_data_buf, allocator, sizeof(key_data) + key->len);
+    aws_byte_buf_write(&key_data_buf, (const uint8_t *)&key_data, sizeof(key_data));
+    aws_byte_buf_write(&key_data_buf, key->buffer, key->len);
+
+    BCRYPT_KEY_HANDLE key_handle;
+    status = BCryptImportKey(
+        algHandle, NULL, BCRYPT_KEY_DATA_BLOB, &key_handle, NULL, 0, key_data_buf.buffer, (ULONG)key_data_buf.len, 0);
+
+    aws_byte_buf_clean_up_secure(&key_data_buf);
+
+    if (!NT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
+
+    return key_handle;
+}
+
+static void s_aes_default_destroy(struct aws_symmetric_cipher *cipher) {
+    struct aes_bcrypt_cipher *cipher_impl = cipher->impl;
+
+    /* If we're in CBC mode the working iv is the same buffer as the iv.
+       so prevent a double free here by checking the addresses first. */
+    bool working_iv_optimized = cipher->iv.buffer == cipher_impl->working_iv.buffer;
+    aws_byte_buf_clean_up_secure(&cipher->key);
+    aws_byte_buf_clean_up_secure(&cipher->iv);
+    aws_byte_buf_clean_up_secure(&cipher->tag);
+    aws_byte_buf_clean_up_secure(&cipher->aad);
+
+    if (!working_iv_optimized) {
+        aws_byte_buf_clean_up_secure(&cipher_impl->working_iv);
```

Comment:
trivial: working_iv has its `allocator` set because it's a by-value copy of `iv`

might be less confusing if the `allocator` wasn't set, when it doesn't actually own its memory